luarules/gadgets/unit_custom_weapons_cluster.lua

```diff
@@ -159,140 +148,212 @@ for wdid, data in pairs(dataTable) do
 
-local unitBulk = {}
 for udid, udef in pairs(UnitDefs) do
-    unitBulk[udid] = min(
-        (   sqrt(udef.health) +
-            sqrt(udef.metalCost) * udef.xsize * udef.zsize * sqrt(udef.radius)
-        ) / minBulkReflect,
-        1.0
-    )
+    -- Set the unit bulk values.
+    -- todo: We don't even _detect_ walls. "Objectified" units aren't returned by GetUnitsInSphere?
+    -- todo: Seems likely that the same goes for wall-like units, then.
+    if udef.armorType == Game.armorTypes.wall or udef.armorType == Game.armorTypes.indestructible then
+        unitBulk[udid] = 0.9
+    elseif udef.name == 'armclaw' or udef.name == 'cormaw' or udef.name == 'legjaw' then
+        unitBulk[udid] = 0.8
+    else
+        unitBulk[udid] = min(
+            1.0,
+            ((  udef.health ^ 0.5 +                         -- HP is log2-ish but that feels too tryhard
+                udef.metalCost ^ 0.5 *                      -- Steel (metal) is heavier than feathers (energy)
+                udef.xsize * udef.zsize * udef.radius ^ 0.5 -- People see 'bigger thing' as 'more solid'
+            ) / minBulkReflect)                             -- Scaled against some large-ish bulk rating
+        ) ^ 0.33                                            -- Raised to a low power to curve up the results
+    end
+end
+
+local bulkMin = unitBulk[UnitDefNames[minUnitBounces].id] or minBulkReflect / 10
+for udid, _ in pairs(UnitDefs) do
+    if unitBulk[udid] < bulkMin then
+        unitBulk[udid] = nil
+    end
 end
 
+-- Spring.Debug.TableEcho(
+--     {
+--         ['Tick bulk']     = unitBulk[ UnitDefNames["armflea"].id  ],
+--         ['Pawn bulk']     = unitBulk[ UnitDefNames["armpw"].id    ],
+--         ['Gauntlet bulk'] = unitBulk[ UnitDefNames["armguard"].id ],
+--         ['Pulsar bulk']   = unitBulk[ UnitDefNames["armanni"].id  ],
+--         ['Thor bulk']     = unitBulk[ UnitDefNames["armthor"].id  ],
+--     }
+-- )
+
+-- Reusable table for reducing garbage
+
+local spawnCache  = {
+    pos     = { 0, 0, 0 },
+    speed   = { 0, 0, 0 },
+    owner   = 0,
+    ttl     = defaultSpawnTtl,
+    gravity = mapGravity,
+}
+
 --------------------------------------------------------------------------------------------------------------
 -- Functions -------------------------------------------------------------------------------------------------
 
-local function RandomVector3()
-    local m1, m2, m3, m4       -- Marsaglia procedure:
-    repeat                     -- The method begins by sampling & rejecting points.
-        m1 = 2 * rand() - 1    -- The result can be transformed into radial coords.
-        m2 = 2 * rand() - 1    -- Rand floats are expensive, though. Might replace.
-        m3 = m1 * m1 + m2 * m2
-    until (m3 < 1)
-    m4 = sqrt(1 - m3)
-    return 2 * m1 * m4 , -- x
-           2 * m2 * m4 , -- y
-           1 -  2 * m3   -- z
+local function RandomVector3(n)
```

Comment:
This and all the math below could be put in a common math library

Reply:
I'm not against that. It has a general use. Random direction vecs compete with engine implementations for random solid angles, like spray angle. I considered using spray angle and projectile count instead, so clusters would work like 360-degree shotguns. But this has a niche use for evenly spacing the projectiles.

Reply:
Moved to a small util.